runtime/jailer.go

```diff
@@ -0,0 +1,299 @@
+			}
+
+			logger.Debugf("Overwritting process args of config")
+			if err := overwriteConfig(cfg, filepath.Base(m.Cfg.SocketPath), rootPathToConfig); err != nil {
+				return errors.Wrap(err, "failed to overwrite config.json")
+			}
+
+			logger.Debugf("chdir to %v", jailPath)
+			if err := os.Chdir(jailPath); err != nil {
+				return errors.Wrapf(err, "failed to chdir into %v", jailPath)
+			}
+
+			// copy the firecracker binary
+			logger.Debugf("copying firecracker binary to %v", binPath)
+			if err := copyFile(cfg.FirecrackerBinaryPath, binPath); err != nil {
+				return errors.Wrapf(err, "could not copy firecracker binary from path %v", cfg.FirecrackerBinaryPath)
+			}
+
+			// Link the kernel image
+			newKernelImagePath := filepath.Join(contentsPath, kernelImageFileName)
+			logger.Debugf("linking kernel image to %v", newKernelImagePath)
+			if err := os.Link(m.Cfg.KernelImagePath, newKernelImagePath); err != nil {
+				return errors.Wrapf(err, "failed to link kernel image from path %v", m.Cfg.KernelImagePath)
+			}
+			m.Cfg.KernelImagePath = filepath.Join("/var/lib/firecracker-containerd", kernelImageFileName)
+
+			logger.Debugf("creating devices in %v", devPath)
+			if err := createDevices(devPath); err != nil {
+				return errors.Wrap(err, "failed to create necessary devices")
+			}
+
+			// link drives to new contents path
+			for i, d := range m.Cfg.Drives {
+				drivePath := firecracker.StringValue(d.PathOnHost)
+				fileName := filepath.Base(drivePath)
+
+				if err := os.Link(drivePath, filepath.Join(contentsPath, fileName)); err != nil {
+					return errors.Wrapf(err, "failed to copy drive %v", drivePath)
+				}
+
+				newDrivePath := filepath.Join(strings.TrimPrefix(contentsPath, rootPath), fileName)
+				m.Cfg.Drives[i].PathOnHost = firecracker.String(newDrivePath)
+			}
+
+			if err := chownr(rootPath, firecracker.IntValue(cfg.UID), firecracker.IntValue(cfg.GID)); err != nil {
+				return errors.Wrapf(err, "failed to chown -r %v", rootPath)
+			}
+
+			logger.Info("Successfully ran jailer handler")
+			return nil
+		},
+	}
+}
+
+// createDevices will create a series of device nodes at the given path
+func createDevices(path string) error {
+	devices := []struct {
+		Name  string
+		Major uint32
+		Minor uint32
+	}{
+		{
+			Name:  "kvm",
+			Major: 10,
+			Minor: 232,
+		},
+		{
+			Name:  "net/tun",
+			Major: 10,
+			Minor: 200,
+		},
+		{
+			Name:  "vhost-vsock",
+			Major: 10,
+			Minor: 241,
+		},
+	}
+
+	for _, d := range devices {
+		devicePath := filepath.Join(path, d.Name)
+		rdev := unix.Mkdev(d.Major, d.Minor)
+		if err := syscall.Mknod(
+			devicePath,
+			// S_IFCHR character special device
+			// S_IRUSR read permission
+			// S_IWUSR write permission
+			syscall.S_IFCHR|syscall.S_IRUSR|syscall.S_IWUSR,
+			int(rdev),
+		); err != nil {
+			return errors.Wrapf(err, "failed to mknode %v", devicePath)
+		}
+	}
+
+	return nil
+}
+
+func copyFile(src, dst string) error {
```

Comment:
Instead of copying files, can we do read-only bind mounts of them? It looks like this is used for the firecracker binary and the runc config.json, both of which shouldn't be writable inside the jail. Read-only bind mounts will prevent any writes (no matter the permissions) and will prevent extra disk space from being used.